shared/naturalcrit/codeEditor/codeEditor.jsx

```diff
@@ -169,9 +146,61 @@ const CodeEditor = createClass({
 		});
 		closeTag.autoCloseCurlyBraces(CodeMirror, this.codeMirror);
 
+		this.setState({
+			widgetUtils : require('./helpers/widgets')(themeWidgets, this.codeMirror, (hints, field)=>{
+				this.setState({
+					hints,
+					hintsField : field
+				});
+			})
+		});
+
 		// Note: codeMirror passes a copy of itself in this callback. cm === this.codeMirror. Either one works.
-		this.codeMirror.on('change', (cm)=>{this.props.onChange(cm.getValue());});
+		this.codeMirror.on('change', (cm)=>{
+			this.props.onChange(cm.getValue());
+			this.state.widgetUtils.updateWidgetGutter();
+		});
+
+		this.codeMirror.on('cursorActivity', (cm)=>{
+			const { line } = cm.getCursor();
+			for (const key in this.state.widgets) {
+				if(key != line) {
+					this.state.widgets[key]?.clear();
+				}
+			}
+			const { widgets } = this.codeMirror.lineInfo(line);
+			if(!widgets) {
+				const widget = this.state.widgetUtils.updateLineWidgets(line);
+				if(widget) {
+					this.setState({
+						widgets : {
+							[line] : widget
+						}
+					});
+				}
+			}
+		});
```

Comment:
Do we need to use keys to track a list of widgets for multiple lines at a time, or can we simply make `this.state.widgets` hold the one currently-selected line?

Seems like the intent is only one line of widgets is open at a time anyway; clicking on a line closes the widgets on all other lines.

I.e., simplify this to:

```
		this.codeMirror.on('cursorActivity', (cm)=>{
			const { line } = cm.getCursor();
			const { widgets } = this.codeMirror.lineInfo(line);

			if(!widgets) {
				this.state.widgets.clear(); //clears the previously-opened line of widgets
				this.setState({
					widgets : this.state.widgetUtils.updateLineWidgets(line) || []
				});
			}
		});
```

Reply:
As things currently stand if you click on the gutter button you can open multiple widgets at a time